### PR TITLE
test: add e2e provider tests

### LINKED ISSUE #2: Add E2E testing
Write e2e tests for Sync controller using `e2e provider` similar to - https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/test/bats/e2e-provider.bats

### PR DESCRIPTION
**Pull request by AlexanderStocks**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR adds end-to-end (E2E) testing for the project, ensuring that the system works as expected through various scenarios.

**Which issue(s) this PR fixes**:
Fixes #2

**Special notes for your reviewer**:
Please review the added BATS tests and ensure they cover the intended scenarios correctly.
**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,6 +15,9 @@ TOOLS_BIN_DIR := $(TOOLS_DIR)/bin
 KUSTOMIZE_VERSION ?= v4.5.7
 CONTROLLER_TOOLS_VERSION ?= v0.11.1
 KIND_NODE_IMAGE_VERSION ?= v1.29.2
+BATS_VERSION ?= 1.11.0
+SHELLCHECK_VER ?= v0.10.0
+KIND_VERSION ?= 0.23.0
 
 ## Tool Binaries
```
Comment (nilekhc) [inline Makefile:22]: add following to invoke install targets

```shell
KIND := kind
ENVSUBST := envsubst
BATS := bats
```

This should fix current CI failures.

**Review by nilekhc — CHANGES_REQUESTED** [review]

**Comment by nilekhc** [comment]
All the review comments from https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/31 have been addressed in the version of the PR.

**Comment by nilekhc** [comment]
Drop [chore: updates kubebuilder config](https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/40/commits/a89b4990063d631398ab724dfe6d950891909a37) commit. You have pulled my commit as well into this PR.

Could also add e2e tests for VAP policies we have? We want to make sure controller is not trying to do things which are restricted using VAP.

**Comment by nilekhc** [comment]
> Drop [chore: updates kubebuilder config](https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/40/commits/a89b4990063d631398ab724dfe6d950891909a37) commit. You have pulled my commit as well into this PR.
> 
> Could also add e2e tests for VAP policies we have? We want to make sure controller is not trying to do things which are restricted using VAP.

VAP e2e tests will be covered by https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/47

**Comment by nilekhc** [comment]
@AlexanderStocks Changes looks good. Could you squash commits into single one?

File: Makefile [hunk 2]
```diff
@@ -40,8 +43,14 @@ CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
 ENVTEST ?= $(LOCALBIN)/setup-envtest
 GOLANGCI_LINT := $(TOOLS_BIN_DIR)/golangci-lint
 HELM := helm
+KIND := kind
+ENVSUBST := envsubst
+BATS := bats
 TRIVY := trivy
 
+# ENVTEST_K8S_VERSION refers to the version of kubebuilder assets to be downloaded by envtest binary.
+ENVTEST_K8S_VERSION = 1.26.0
```
Comment (nilekhc) [inline Makefile:52]: Is this required? Remove if not.

**Review by nilekhc — CHANGES_REQUESTED** [review]

File: test/bats/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Testing Secrets Store Sync Controller with e2e provider
+
+This directory contains e2e test scripts for the Secrets Store Sync Controller. Before running the e2e tests, install
+the Secrets Store Sync Controller with the `e2e provider` as outlined as outlined [here](../../README.md#getting-started).
```
Comment (aramase) [inline test/bats/README.md:4]:
```suggestion
the Secrets Store Sync Controller with the `e2e provider` as outlined [here](../../README.md#getting-started).
```

File: test/bats/e2e-provider.bats [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+  run kubectl get clusterrolebinding/secrets-store-sync-controller-manager-rolebinding
+  assert_success 
+}
+
+@test "[v1alpha1] deploy e2e-providerspc secretproviderclass crd" {
+  kubectl create namespace test-v1alpha1 --dry-run=client -o yaml | kubectl apply -f -
+
+  kubectl apply -n test-v1alpha1 -f $BATS_RESOURCE_MANIFESTS_DIR/e2e-providerspc.yaml
+  kubectl wait --for condition=established -n test-v1alpha1 --timeout=60s crd/secretproviderclasses.secrets-store.csi.x-k8s.io
```
Comment (aramase) [inline test/bats/e2e-provider.bats:36]: This is already checked [here](https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/40/files#diff-395e5ad96d74d81648ec476082473348247fb32c3bbe242149b3c8b4144e7d85R11), so not required.

File: test/bats/e2e-provider.bats [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+@test "Test rbac roles and role bindings exist" {
+  run kubectl get clusterrole/secrets-store-sync-controller-manager-role
+  assert_success
+
+  run kubectl get clusterrolebinding/secrets-store-sync-controller-manager-rolebinding
+  assert_success 
+}
+
+@test "[v1alpha1] deploy e2e-providerspc secretproviderclass crd" {
```
Comment (aramase) [inline test/bats/e2e-provider.bats:32]: What is the purpose of this test? We're doing the same set of steps again [here](https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/40/files#diff-395e5ad96d74d81648ec476082473348247fb32c3bbe242149b3c8b4144e7d85R43-R48).

File: test/bats/e2e-provider.bats [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  # Create the SecretSync
+  kubectl apply -n test-v1alpha1 -f $BATS_RESOURCE_MANIFESTS_DIR/e2e-secret-sync.yaml
+  kubectl wait --for condition=established -n test-v1alpha1 --timeout=60s crd/secretsyncs.secret-sync.x-k8s.io
+
+  cmd="kubectl get secretsyncs.secret-sync.x-k8s.io/sse2esecret -n test-v1alpha1 -o yaml | grep sse2esecret"
+  wait_for_process $WAIT_TIME $SLEEP_TIME "$cmd"
+
+  # Retrieve the secret 
+  cmd="kubectl get secret sse2esecret -n test-v1alpha1 -o yaml | grep 'apiVersion: secret-sync.x-k8s.io/v1alpha1'"
+  wait_for_process $WAIT_TIME $SLEEP_TIME "$cmd"
```
Comment (aramase) [inline test/bats/e2e-provider.bats:59]: In addition to this, need to check the data in the secret matches what we expect

File: test/bats/e2e-provider.bats [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+  # Delete the secret
+  kubectl delete secretsync sse2esecret -n test-v1alpha1
```
Comment (aramase) [inline test/bats/e2e-provider.bats:62]: After this step, check if the Kubernetes secret has been deleted

File: test/bats/e2e-provider.bats [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  cmd="kubectl get secretsyncs.secret-sync.x-k8s.io/sse2esecret -n ss-namespace -o yaml | grep sse2esecret"
+  wait_for_process $WAIT_TIME $SLEEP_TIME "$cmd"
+
+  # Check that the secret is not created in ss-namespace
+  cmd="kubectl get secret sse2esecret -n ss-namespace"
+  run $cmd
+  assert_failure
+}
+
+@test "Delete SecretSync and check that the secret is deleted" {
```
Comment (aramase) [inline test/bats/e2e-provider.bats:90]: Added a comment above. I don't think we need a separate test for this. I would recommend validating every required step in `@test "[v1alpha1] deploy e2e-providerspc secretsync crd"` as a sequence.

File: test/bats/helpers.bash
Comment (aramase) [inline test/bats/helpers.bash:1]: Need to add log collection as part of the helper, so we have logs for each test run

**Review by aramase — CHANGES_REQUESTED** [review]

**Review by aramase — CHANGES_REQUESTED** [review]
Almost there. Just few more comments.

**Review by aramase — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**